content/blog/2025-12-07-avoid-consecutive-repartitions.md

```diff
@@ -0,0 +1,428 @@
+
+### The Root Cause
+
+With a single rule to read, isolating the issue is much simpler. The EnforceDistribution rule takes a physical query plan as input, iterates over each child analyzing its requirements, and decides where adding repartition nodes is beneficial.
+
+A great place to start looking is before any repartitions are inserted, and where the program decides if adding a repartition above/below an operator is useful. With the help of handy function header comments, it was easy to identify that this is done in the [get_repartition_requirement_status](https://github.com/apache/datafusion/blob/944f7f2f2739a9d82ac66c330ea32a9c7479ee8b/datafusion/physical-optimizer/src/enforce_distribution.rs#L1108) function. Here, Datafusion sets four fields indicating how the operator would benefit from repartitioning:
+
+1. **The operator's distribution requirement**: what type of partitioning does it need from its children (hash, single, or unknown)?
+2. **If round-robin is theoretically beneficial:** does the operator benefit from parallelism?
+3. **If our data indicates round-robin to be beneficial**: do we have enough data to justify the overhead of repartitioning?
+4. **If hash repartitioning is necessary**: is the parent an operator that requires all column values to be in the same partition, like an aggregate, and are we already hash-partitioned correctly?
+
+Ok, great\! We understand the different components Datafusion uses to indicate if repartitioning is beneficial. Now all that's left to do is see how repartitions are inserted.
+
+This logic takes place in the main loop of this rule. I find it helpful to draw algorithms like these into logic trees; this tends to make things much more straightforward and approachable:
+
+<div class="text-center">
+<img
+  src="/blog/images/avoid-consecutive-repartitions/logic_tree_before.png"
+  width="75%"
+  class="img-responsive"
+  alt="Incorrect Logic Tree"
+/>
+</div>
+<br>
+
+Boom\! This is the root of our problem: we are inserting a round-robin repartition, then still inserting a hash repartition afterwards. This means that if an operator indicates it would benefit from both round-robin and hash repartitioning, consecutive repartitions will occur.
+
+---
+
+## **The Fix**
+
+The logic shown before is, of course, incorrect, and the conditions for adding hash and round-robin repartitioning should be mutually exclusive since an operator will never benefit from shuffling data twice.
+
+Well, what is the correct logic?
+
+Based on our lesson on hash repartitioning and the indicators Datafusion uses to determine when repartitioning can benefit an operator, the fix is easy. In the sub-tree where an operator's parent requires hash partitioning:
```

Comment:
nit: "heuristics" is another term for this type of rule -- maybe this would read better

```suggestion
Based on our lesson on hash repartitioning and the heuristics Datafusion uses to determine when repartitioning can benefit an operator, the fix is easy. In the sub-tree where an operator's parent requires hash partitioning:
```